control-plane/typeSpecConverter.cpp

```diff
@@ -0,0 +1,243 @@
+        headerUnionStackTypeSpec->mutable_header_union()->set_name(name);
+        headerUnionStackTypeSpec->set_size(size);
+    } else {
+        BUG("Unexpected declaration %1%", decl);
+    }
+    visit(decl->getNode());
+    map.emplace(type, typeSpec);
+    return false;
+}
+
+bool TypeSpecConverter::preorder(const IR::Type_Struct* type) {
+    if (p4RtTypeInfo) {
+        auto name = std::string(type->controlPlaneName());
+        auto structs = p4RtTypeInfo->mutable_structs();
+        if (structs->find(name) == structs->end()) {
+            auto structTypeSpec = new p4::P4StructTypeSpec();
+            for (auto f : type->fields) {
+                auto fType = f->type;
+                visit(fType);
+                auto fTypeSpec = map.at(fType);
+                auto member = structTypeSpec->add_members();
+                member->set_name(f->controlPlaneName());
+                member->mutable_type_spec()->CopyFrom(*fTypeSpec);
+            }
+            (*structs)[name] = *structTypeSpec;
+        }
+    }
+    map.emplace(type, nullptr);
+    return false;
+}
+
+bool TypeSpecConverter::preorder(const IR::Type_Header* type) {
+    if (p4RtTypeInfo) {
+        auto name = std::string(type->controlPlaneName());
+        auto headers = p4RtTypeInfo->mutable_headers();
+        if (headers->find(name) == headers->end()) {
+            auto headerTypeSpec = new p4::P4HeaderTypeSpec();
+            for (auto f : type->fields) {
+                auto fType = f->type;
+                visit(fType);
+                auto fTypeSpec = map.at(fType);
+                BUG_CHECK(fTypeSpec->has_bitstring(),
+                          "Only bistring fields expected in header type declaration %1%", type);
+                auto member = headerTypeSpec->add_members();
+                member->set_name(f->controlPlaneName());
+                member->mutable_type_spec()->CopyFrom(fTypeSpec->bitstring());
+            }
+            (*headers)[name] = *headerTypeSpec;
+        }
+    }
+    map.emplace(type, nullptr);
+    return false;
+}
+
+bool TypeSpecConverter::preorder(const IR::Type_HeaderUnion* type) {
+    if (p4RtTypeInfo) {
+        auto name = std::string(type->controlPlaneName());
+        auto headerUnions = p4RtTypeInfo->mutable_header_unions();
+        if (headerUnions->find(name) == headerUnions->end()) {
+            auto headerUnionTypeSpec = new p4::P4HeaderUnionTypeSpec();
+            for (auto f : type->fields) {
+                auto fType = f->type;
+                visit(fType);
+                auto fTypeSpec = map.at(fType);
+                BUG_CHECK(fTypeSpec->has_header(),
+                          "Only header fields expected in header union declaration %1%", type);
+                auto member = headerUnionTypeSpec->add_members();
+                member->set_name(f->controlPlaneName());
+                member->mutable_header()->CopyFrom(fTypeSpec->header());
```

Comment:
Are header_unions just as structs?

Reply:
They are similar enough that it's tempting to unify the code, yet different enough that I couldn't find a good way to do it...

Reply:
Thanks! I can see the extra checks for members of a header union are only headers.